tooi/api/streaming.py

```diff
@@ -0,0 +1,241 @@
+                    f"{str(exc)}"))
+                await asyncio.sleep(self.ERROR_BACKOFF)
+
+    async def _stream(self):
+        logger.info("HTTPStreamClient: connecting to stream={stream}")
+
+        async with self.ctx.auth.client.stream("GET", self.url, timeout=self.TIMEOUT) as stream:
+            await self._parse_stream(stream)
+
+    async def _parse_stream(self, stream):
+        async for line in stream.aiter_lines():
+            await self._handle_line(line)
+
+    async def _handle_line(self, line: str):
+        # Lines beginning with ':' are comments (keepalives).
+        if len(line) == 0 or line[0] == ':':
+            return
+
+        # Events are two lines long.  If we haven't get the second line yet, buffer this one and try
+        # again next time.
+        if len(self._lines) == 0:
+            self._lines.append(line)
+            return
+
+        # We should have both lines now.
+        event_line = self._lines[0]
+        data_line = line
+        self._lines = []
+
+        if (m := re.match(r"^event:\s*(.*)$", event_line)) is None:
+            # TODO: Disconnect or log error?
+            return
+
+        event_type = m.group(1)
+
+        if (m := re.match(r"^data:\s*(.*)$", data_line)) is None:
+            # TODO: Disconnect or log error?
+            return
+
+        event_json = m.group(1)
+
+        try:
+            event_payload = json.loads(event_json)
+        except json.JSONDecodeError:
+            logger.info("HTTPStreamClient: could not decode JSON")
+            return
+
+        event = StreamEvent(self.stream_name, event_type, event_payload)
+        logger.info(f"HTTPStreamClient: push event={event_type} to queue for {self.stream_name}")
+        await self.queue.put(event)
+
+
+class StreamSubscription(object):
+    # Stream types
+    USER = "user"
+    PUBLIC_LOCAL = "public/local"
+    PUBLIC_REMOTE = "public/remote"
+
+    def __init__(self, mplx: "StreamMultiplexer", stream: str):
+        self.mplx = mplx
+        self.stream = stream
+        self.queue = StreamQueue()
+
+    async def get(self):
+        e = await self.queue.get()
+        return e
+
+    async def dispatch(self, e: StreamEvent):
+        await self.queue.put(e)
+
+    async def close(self):
+        await self.mplx.close_stream(self)
+        self.mplx = None
+
+
+class StreamMultiplexer(object):
+    """
+    Internal class for multiplexing multiple streams into a single queue.  Call open_stream() to
+    start receiving events on a stream, and close_stream() to stop receiving them.  The number of
+    listeners for each stream type is tracked, and the stream won't be closed unless close_stream()
+    is called the same number of times open_stream() was called.
+    """
+
+    # lock order: self.lock -> StreamInstance.lock
+    #         or: StreamInstance.lock if self.lock not held
+    #        NOT: StreamInstance.lock -> self.lock
+
+    class StreamInstance(object):
```

Comment:
Can this class not be nested? I don't see any advantages.

Reply:
fixed in 4490988977125d6815145d9f255e2168a2287990